### PR TITLE
Feature: two-factor authentication

### PR DESCRIPTION
**Pull request by shamoon**
## Proposed change

This pretty much entirely uses allauth's handling of totp and just kinda connects it to our frontend.

Screenshot 2024-10-25 at 10 55 34 AM
Screenshot 2024-10-25 at 11 01 26 AM
Screenshot 2024-10-25 at 11 01 52 AM
Screenshot 2024-10-25 at 10 55 09 AM

Closes #424

## Type of change

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by chwiese** [comment]
Thank you for working on this @shamoon. I saw that [allauth supports WebAuthn (Passkeys) for MFA](https://django-allauth.readthedocs.io/en/latest/mfa/webauthn.html). Would you consider supporting that in addition to TOTP please?

File: src/paperless/views.py [hunk 1]
```diff
@@ -145,6 +171,76 @@ def patch(self, request, *args, **kwargs):
...
+        user = self.request.user
+        authenticator = Authenticator.objects.filter(
+            user=user,
+            type=Authenticator.Type.TOTP,
+        ).first()
+        if authenticator is not None:
+            delete_and_cleanup(request, authenticator)
+            return Response(True)
+        else:
+            return HttpResponseBadRequest("TOTP not found")
```
Comment (stumpylog) [inline src/paperless/views.py:241]: Not sure, but if it is not found, maybe HttpResponseNotFound is a little more precise?

File: src/paperless/views.py [hunk 2]
```diff
@@ -100,6 +108,24 @@ class UserViewSet(ModelViewSet):
...
+            )
+        authenticator = Authenticator.objects.filter(
+            user=user,
+            type=Authenticator.Type.TOTP,
+        ).first()
+        if authenticator is not None:
+            delete_and_cleanup(request, authenticator)
+            return Response(True)
+        else:
+            return HttpResponseBadRequest("TOTP not found")
```
Comment (stumpylog) [inline src/paperless/views.py:127]: Think about it here too

**Review by stumpylog — COMMENTED** [review]
Need to test it out too, but a couple minor thoughts

File: src/paperless/views.py (hunk 1 above)
Comment (shamoon) [reply src/paperless/views.py:241]: Sure x2!

**Comment by stumpylog** [comment]
Did some testing, all seems to work.  Anyway we can customize the name?  Mine came across as `root@localhost`.  Maybe use the username of the user?

**Comment by shamoon** [comment]
Hmm, Im not certain where the name comes from, my 2FA app doesn't really show that.

I did try setting the 'issuer': https://github.com/paperless-ngx/paperless-ngx/blob/f80e85ef3ba4c0378ce9968c1487d70ed829d0b5/src/paperless/settings.py#L462 and it seemed to get the email correctly for me, so for example mine comes out as: `otpauth://totp/me%40example.com?secret=PQQ73YEPH3K3QACWY4DHDOZJWMLID5UP&issuer=Paperless-ngx`

Is `root@localhost` your email you were testing with?

**Review by stumpylog — APPROVED** [review]
Ah, that's the source.  I don't set the email when I auto create a user

**Pull request merged**